src/language-js/parser-babylon.js

```diff
@@ -38,35 +37,49 @@ function babelOptions(extraOptions, extraPlugins) {
         "throwExpressions",
         "logicalAssignment",
         "classPrivateMethods"
-      ].concat(extraPlugins)
+      ].concat(extraPlugins, includeJsx ? "jsx" : [])
     },
     extraOptions
   );
 }
 
+function createCombinations(extraPlugins, includeJsx) {
+  return [
+    babelOptions(
+      { strictMode: true },
+      ["decorators-legacy"].concat(extraPlugins),
+      includeJsx
+    ),
+    babelOptions(
+      { strictMode: false },
+      ["decorators-legacy"].concat(extraPlugins),
+      includeJsx
+    ),
+    babelOptions(
+      { strictMode: true },
+      [["decorators", { decoratorsBeforeExport: false }]].concat(extraPlugins),
+      includeJsx
+    ),
+    babelOptions(
+      { strictMode: false },
+      [["decorators", { decoratorsBeforeExport: false }]].concat(extraPlugins),
+      includeJsx
+    )
+  ];
+}
+
 function createParse(parseMethod, extraPlugins) {
   return (text, parsers, opts) => {
     // Inline the require to avoid loading all the JS if we don't use it
     const babel = require("@babel/parser");
 
-    const combinations = [
-      babelOptions(
-        { strictMode: true },
-        ["decorators-legacy"].concat(extraPlugins)
-      ),
-      babelOptions(
-        { strictMode: false },
-        ["decorators-legacy"].concat(extraPlugins)
-      ),
-      babelOptions(
-        { strictMode: true },
-        [["decorators", { decoratorsBeforeExport: false }]].concat(extraPlugins)
-      ),
-      babelOptions(
-        { strictMode: false },
-        [["decorators", { decoratorsBeforeExport: false }]].concat(extraPlugins)
-      )
-    ];
+    let combinations = createCombinations(extraPlugins);
+
+    if (extraPlugins && extraPlugins.indexOf("typescript") !== -1) {
+      combinations = combinations.concat(
+        createCombinations(extraPlugins, false)
```

Comment:
This isn't good. The current `combinations` is already a hack. If we explode this number of combinations, this might lead to a very noticeable performance degradation

Reply:
You mean we should check `extraPlugins.indexOf("typescript") !== -1` first to avoid double `combinations` size?

Reply:
It could be optimized by only enable `decorators-legacy` plugin for `typescript`, I will refactor it ASAP.

Reply:
@duailibe @lipis Done.

Reply:
@duailibe Can you help to review again?